foo/lib/config/__init__.py

```diff
@@ -0,0 +1,3 @@
+from .configloader import ConfigLoader, ConfigLoaderException
```

Comment:
IMO Singleton pattern adds a lot of overhead with regard to code (overriding `__new__`, keeping track of state) and tests. How about using Strategy pattern?

If we really want to restrict loading config only once - we can use `Loader` class as an interface.

```python

from abc import abstractmethod, ABC

import logging

logger = logging.getLogger(__name__)


class BaseLoadingStrategy(ABC):

    def __init__(self, defaults: dict):
        self.defaults = defaults

    def load(self, path: str) -> Config:
        config = {
            **self.defaults,
            **self._load_external(path)
        }
        environment_config = self._load_environment_config(config)
        config.update(environment_config)
        return Config(config, path)

    @abstractmethod
    def _load_external(self, path: str) -> dict:
        pass

    @staticmethod
    def _load_environment_config(config: dict) -> dict:
        environ = os.environ
        return {key: environ[key] for key in config if key in environ}


class LocalLoadingStrategy(BaseLoadingStrategy):

    def _load_external(self, path: str) -> dict:
        try:
            with open(path) as f:
                return yaml.load(f)
        except FileNotFoundError as e:
            raise ConfigLoaderException(f'Error opening {path}') from e


class S3LoadingStrategy(BaseLoadingStrategy):
    S3_BUCKET = 'configs'

    def _load_external(self, path: str) -> dict:
        try:
            f = boto3.client('s3').download_fileobj(self.S3_BUCKET, path)
            return yaml.load(f)
        except Exception:
            raise ConfigLoaderException(f'Error when opening {path}')


class Loader:
    CONFIG: Config = None
    _STRATEGIES = {
        'test': LocalLoadingStrategy,
        'prod': S3LoadingStrategy
    }

    @classmethod
    def load(cls, path: str, defaults: dict = None) -> Config:
        return cls.CONFIG if cls.CONFIG else cls._load(defaults or {}, path)

    @classmethod
    def _load(cls, defaults: dict, path: str) -> Config:
        env = cls._detect_environment()
        cls.CONFIG = cls._STRATEGIES[env](defaults).load(path)
        return cls.CONFIG

    @classmethod
    def _detect_environment(cls) -> str:
        with open('env') as file:
            env = file.read().strip()
            assert env in cls._STRATEGIES
            logger.info(f'Current environment is: {env!r}')
            return env
```